File: leetcode/backspace_string/backspace_string.cpp
#include<iostream>
#include<regex>
using namespace std;

string applyBackspace(string S)
{
  auto sub = S.find("#");
  if (sub == string::npos) return S;
  regex re("[a-z]#");
  return applyBackspace(regex_replace(S, re, ""));
}

bool backspaceCompare(string S, string T)
{
  if (S.length() != T.length()) return false;
  string sb = applyBackspace(S), tb = applyBackspace(T);
  cout << sb << " " << tb << endl;
  return sb == tb;
}

int main(int argc, char *argv[])
{
  cout << backspaceCompare("ab#c", "ad#c") << endl;
  cout << backspaceCompare("ab##", "c#d#") << endl;
  cout << backspaceCompare("a##c", "#a#c") << endl;
  cout << backspaceCompare("a#c", "b") << endl;
  return 0;
}

File: leetcode/backspace_string/backspace_string.go
package main

import (
	"fmt"
)

func applyBackspace(word string) string {
	pos := -1
	for i, c := range word {
		if c == '#' {
			pos = i
			break
		}
	}
	if pos == -1 {
		return word
	}
	if pos == 0 {
		word = word[pos+1:]
	} else {
		word = word[:pos-1] + word[pos+1:]
	}
	return applyBackspace(word)
}

func backspaceCompare(S string, T string) bool {
	return applyBackspace(S) == applyBackspace(T)
}

func main() {
	fmt.Println(backspaceCompare("ab#c", "ad#c"))
	fmt.Println(backspaceCompare("ab##", "c#d#"))
	fmt.Println(backspaceCompare("a##c", "#a#c"))
	fmt.Println(backspaceCompare("a#c", "b"))
}
